src/main/java/vct/col/rewrite/InlinePredicatesRewriter.java

```diff
@@ -68,43 +84,97 @@ public void visit(Method m){
+   * is not at the bottom of a multiplication tree, some type error is triggered in Viper. Ideally,
+   * this method would have no arguments, and would just return the multiplication tree of the scalars.
+   *
+   * @param startNode Node to start the multiplication with. If not available, can be null.
+   * @return AST corresponding to: (((startNode * scalar1) * scalar2) * ...
+   */
+  private ASTNode getCombinedScalar(ASTNode startNode) {
+    if (inlinedScalars.empty()) {
+      Abort("Cannot combine scalars when stack is empty");
+    }
+
+    ASTNode result;
+    int startIndex;
+
+    if (startNode == null) {
+      result = create.local_name(inlinedScalars.get(0));
+      startIndex = 1;
+    } else {
+      result = startNode;
+      startIndex = 0;
+    }
+
+    for (int i = startIndex; i < inlinedScalars.size(); i++) {
+      result = create.expression(
+              StandardOperator.Mult,
+              result,
+              create.local_name(inlinedScalars.get(i)));
+    }
+
+    return result;
+  }
+
   @Override
   public void visit(OperatorExpression e){
     switch(e.operator()){
-      case Unfolding:
-      {
-        ASTNode arg1=rewrite(e.arg(0));
-        ASTNode arg2=rewrite(e.arg(1));
-        if (arg1 instanceof MethodInvokation || arg1.isa(StandardOperator.Scale)){
-          result=create.expression(StandardOperator.Unfolding,arg1,arg2);
+      case Scale:
+        if (inline(e)) {
+          String scaleAmountName = "inlineScalar" + count++;
+
+          inlinedScalars.push(scaleAmountName);
+          super.visit(e);
+          OperatorExpression scaleExpr = (OperatorExpression) result;
+          inlinedScalars.pop();
+
+          result = create.let_expr(
+                  create.field_decl(scaleAmountName, rewrite(e.arg(0).getType()), scaleExpr.arg(0)),
+                  scaleExpr.arg(1)
+          );
         } else {
-          result=arg2;
+          super.visit(e);
         }
         break;
-      }
-      default:
+      case Perm:
         super.visit(e);
-        break;
-    }
-  }
-  @Override
-  public void visit(ASTSpecial e){
-    switch(e.kind){
-      case Unfold:
-      case Fold:
-      { 
-        ASTNode arg=rewrite(e.getArg(0));
-        if (arg instanceof MethodInvokation || arg.isa(StandardOperator.Scale)){
-          result=create.special(e.kind,arg);
-        } else {
-          result=null; // returning null for a statement means already inserted or omit.
+        OperatorExpression newPerm = (OperatorExpression) result;
+        if (!inlinedScalars.empty()) {
+          ASTNode permissionAmount = getCombinedScalar(e.arg(1));
+          result = create.expression(StandardOperator.Perm, newPerm.arg(0), permissionAmount);
         }
         break;
-      }
       default:
         super.visit(e);
         break;
     }
   }
+
+  @Override
+  public void visit(ASTSpecial e) {
+      switch (e.kind) {
+        case Fold:
+        case Unfold:
+          if (inline(e.getArg(0))) {
```

Comment:
Still think we should discuss this in the next meeting :)

Reply:
Yes it is in my notes! ;)